File: Sources/ExcalidrawGeometry/Binding.swift
import ExcalidrawMath
import ExcalidrawModel
import Foundation

/// Arrow-to-shape binding geometry (`packages/element/src/binding.ts`,
/// simplified). Finds the bindable element near an arrow endpoint and converts
/// between an endpoint and a fixed-point ratio within an element's bounds.
public enum Binding {
    /// Default search distance around an element's bounds, in scene units.
    public static let distance = 16.0

    public static func isBindable(_ element: ExcalidrawElement) -> Bool {
        switch element.kind {
        case .rectangle, .diamond, .ellipse, .text, .image, .frame, .magicframe, .embeddable, .iframe:
            true
        case .arrow, .line, .freedraw, .selection:
            false
        }
    }

    /// The smallest bindable element whose bounds (expanded by `threshold`)
    /// contains `point`, excluding `excluding`.
    public static func bindableElement(
        at point: Point, in elements: [ExcalidrawElement], excluding: Set<String>, threshold: Double = distance
    ) -> ExcalidrawElement? {
        var best: (element: ExcalidrawElement, area: Double)?
        for element in elements where isBindable(element) && !excluding.contains(element.id) {
            let b = ElementGeometry.bounds(element)
            let expanded = BoundingBox(
                minX: b.minX - threshold, minY: b.minY - threshold,
                maxX: b.maxX + threshold, maxY: b.maxY + threshold
            )
            if expanded.contains(point) {
                let area = b.width * b.height
                if best == nil || area < best!.area { best = (element, area) }
            }
        }
        return best?.element
    }

    /// Ratio (0–1 on each axis) of `point` within `bounds`.
    public static func fixedPoint(for point: Point, in bounds: BoundingBox) -> Point {
        Point(
            bounds.width == 0 ? 0.5 : (point.x - bounds.minX) / bounds.width,
            bounds.height == 0 ? 0.5 : (point.y - bounds.minY) / bounds.height
        )
    }

    /// The scene point for a fixed-point ratio within `bounds`.
    public static func point(forFixedPoint fixed: Point, in bounds: BoundingBox) -> Point {
        Point(bounds.minX + fixed.x * bounds.width, bounds.minY + fixed.y * bounds.height)
    }
}

File: web/packages/geometry/src/binding.ts
import { Point } from "@cyberdynecorpai/math";
import type { ExcalidrawElement } from "@cyberdynecorpai/model";
import { BoundingBox } from "./bounding-box.js";
import { bounds } from "./element-geometry.js";

/** Default search distance around an element's bounds, in scene units. */
export const BINDING_DISTANCE = 16;

export function isBindable(el: ExcalidrawElement): boolean {
  switch (el.type) {
    case "rectangle":
    case "diamond":
    case "ellipse":
    case "text":
    case "image":
    case "frame":
    case "magicframe":
    case "embeddable":
    case "iframe":
      return true;
    default:
      return false;
  }
}

/**
 * The smallest bindable element whose bounds (expanded by `threshold`) contains
 * `point`, excluding `excluding`. (parity: Binding.swift)
 */
export function bindableElementAt(
  point: Point,
  elements: ExcalidrawElement[],
  excluding: Set<string>,
  threshold = BINDING_DISTANCE,
): ExcalidrawElement | null {
  let best: { element: ExcalidrawElement; area: number } | null = null;
  for (const el of elements) {
    if (!isBindable(el) || excluding.has(el.id)) continue;
    const b = bounds(el);
    const expanded = new BoundingBox(
      b.minX - threshold,
      b.minY - threshold,
      b.maxX + threshold,
      b.maxY + threshold,
    );
    if (expanded.contains(point)) {
      const area = b.width * b.height;
      if (best === null || area < best.area) best = { element: el, area };
    }
  }
  return best?.element ?? null;
}

/** Ratio (0–1 on each axis) of `point` within `bounds`. */
export function fixedPointFor(point: Point, box: BoundingBox): Point {
  return new Point(
    box.width === 0 ? 0.5 : (point.x - box.minX) / box.width,
    box.height === 0 ? 0.5 : (point.y - box.minY) / box.height,
  );
}

/** The scene point for a fixed-point ratio within `bounds`. */
export function pointForFixedPoint(fixed: Point, box: BoundingBox): Point {
  return new Point(box.minX + fixed.x * box.width, box.minY + fixed.y * box.height);
}
